OpenRA.Mods.Common/SpriteLoaders/GifLoader.cs

```diff
@@ -0,0 +1,308 @@
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OpenRA.Mods.Common.SpriteLoaders
+{
+	class GifLoader : ISpriteLoader
+	{
+		public class GifFrame : ISpriteFrame
+		{
+			public Size Size { get; private set; }
+			public Size FrameSize { get; private set; }
+			public float2 Offset { get { return offset; } }
+
+			public byte[] Data { get; set; }
+			public bool DisableExportPadding { get { return false; } }
+
+			float2 offset;
+
+			public GifFrame(Stream s, Size size)
+			{
+				var imageLeft = s.ReadUInt16();
+				var imageTop = s.ReadUInt16();
+				var imageWidth = s.ReadUInt16();
+				var imageHeight = s.ReadUInt16();
+
+				var unpacked = s.ReadUInt8();
+
+				// Local palette
+				if (GetBitValue(unpacked, 0) != 0)
+				{
+					var bits = new byte[]{
+						GetBitValue(unpacked, 5),
+						GetBitValue(unpacked, 6),
+						GetBitValue(unpacked, 7)
+					};
+					
+					s.Position += 3 * (1 << (BitsToNum(bits) + 1));
+				}
+
+				var interlaced = false;
+
+				// Interlaced
+				if (GetBitValue(unpacked, 1) != 0)
+					interlaced = true;
+
+				var lzwMinCodeSize = s.ReadUInt8();
+
+				var data = ReadSubblock(s);
+				data = Decode(lzwMinCodeSize, data);
+
+				if (interlaced)
+				{
+					var originData = data;
+
+					var offsets = new int[]{0, 4, 2, 1};
+					var steps = new int[]{8, 8, 4, 2};
+
+					var j = 0;
+					for (var pass = 0; pass < 4; pass++)
+					{
+						for (var i = offsets[pass]; i < originData.Length; i += steps[pass])
+						{
+							data[i] = originData[j];
+							j++;
+						}
+					}
+				}
+
+				Size = size;
+				FrameSize = new Size(imageWidth, imageHeight);
+				this.offset = new float2(imageLeft, imageTop);
+
+				Data = data;
+			}
+		}
+
+		bool IsGif(Stream s)
+		{
+			var start = s.Position;
+
+			var a = s.ReadASCII(6);
+
+			s.Position = start;
+			return a.StartsWith("GIF");
+		}
+
+		GifFrame[] ParseFrames(Stream s)
+		{
+			var start = s.Position;
+			s.Position += 3;
+			var version = s.ReadASCII(3);
+			if (version != "87a" && version != "89a")
+				throw new Exception("Gif: Not a valid version");
+
+			var width = s.ReadUInt16();
+			var height = s.ReadUInt16();
+			var size = new Size(width, height);
+
+			var GCT = s.ReadUInt8();
+			s.Position += 2;
+
+			if (GetBitValue(GCT, 0) != 0)
+			{
+				var GCEntries = 1 << ((GCT & 0x7) + 1);
+
+				// Skip over palette.
+				s.Position += GCEntries * 3;
+			}
+
+			var frames = new List<GifFrame>();
+
+			while (true)
+			{
+				var block = s.ReadUInt8();
+
+				switch ((BlockTypes)block)
+				{
+					case BlockTypes.ExtensionBlock:
+						var type = (BlockTypes)s.ReadUInt8();
+
+						// We dont care about any of these, so we skip them.
+						switch (type)
+						{
+							case BlockTypes.GraphicExtensionBlock:
+								var lengthG = s.ReadUInt8();
+								if (lengthG == 4)
+									s.Position += 5;
+								else
+									s.Position -= 2;
+								break;
+							case BlockTypes.CommentExtensionBlock:
+								ReadSubblock(s);
```

Comment:
Creating a List, reading N bytes into it, and then discarding that is inefficient.  Just read the count and then advance the position (here and below).